closure.ml

```diff
@@ -19,19 +25,54 @@ and closure_stmt envs statement = match statement with
+and location_and_type envs s = (match envs with 
+  local :: prev_and_global -> if StringMap.mem s local
+                              then (true, StringMap.find s local)
+                              else let rec inclusion envlist = match envlist with
+                                    global :: [] -> if StringMap.mem s global
+                                      then (true, StringMap.find s global)
+                                      else raise (Failure ("Unbound Identifier " ^ s))
+                                  | prev :: rest -> if StringMap.mem s prev
+                                                                then (false, StringMap.find s prev)
+                                                                else inclusion rest
+                                  
+                                  | _ -> raise (Failure "Wrong sized environments")
+                                  in inclusion prev_and_global
+  | _ -> raise (Failure "Wrong sized environments"))
 and closure_expr envs expression = match expression with
     SLiteral(_)      -> []
   | SFliteral(_)     -> []
   | SBoolLit(_)      -> []
-  | SId(_)           -> raise (Failure "Not implemented")
-  | SBinop(_)        -> raise (Failure "Not implemented")
-  | SUnop(_)         -> raise (Failure "Not implemented")
-  | SAssign(_)       -> raise (Failure "Not implemented")
-  | SAssignList(_)   -> raise (Failure "Not implemented")
-  | SCall(_)         -> raise (Failure "Not implemented")
-  | SRecordAccess(_) -> raise (Failure "Not implemented")
-  | SLambda(_)       -> raise (Failure "Not implemented")
-  | SNoexpr          -> []
\ No newline at end of file
+  | SId(s)           -> let (loc, typ) = location_and_type envs s in
+                        if loc
+                        then []
+                        else [(typ, s)]
+  | SBinop((_, e1), _, (_, e2))   -> (union (closure_expr envs e1) (closure_expr envs e2))
+  | SUnop(_, (_, e))         -> (closure_expr envs e)
+  | SAssign(e1, e2)       -> (match e1 with
+    (_, SId(s)) -> let (loc, _) = location_and_type envs s in
+                   if loc
+                   then collect_closure_expr envs (e1::[e2])
+                   else raise (Failure ("Attempt to reassign to \"" ^ s ^ "\" from closure"))
+    | _ -> collect_closure_expr envs (e1::[e2]))
+  | SAssignList(_)   -> raise (Failure "Not implemented5")
+  | SCall(e, args) -> collect_closure_expr envs (e::args)
+                       (* For each argument, compute closure for expression.  Then, combine the closures with union *)
+                       (* Cannot currently test against ID calls other than putChar *)
+  | SRecordAccess(_) -> raise (Failure "Not implemented7")
+  | SLambda(l) -> let (locals::_) = envs in 
+  (*Returns the list l_super minus any elements that are in the StringMap m_sub*)
+                  let rec diff l_super m_sub = (match l_super with
+                      ((t, s)::tail) -> (try if t = StringMap.find s m_sub
```

Comment:
Is there ever a situation in which the type in the inner lambda's closure would not be equal to the type of the outer lambda's local variables? I can't think of a situation but I might be wrong.